File: scripts/ble-scan.swift
// BLE advertisement scanner via CoreBluetooth (no dependencies, uses system Swift).
// Prints every BLE device seen advertising during the scan window.
//
// Build & run:
//   swift scripts/ble-scan.swift 25
// While scanning, hold MO(1)+Q on the NEO65 to trigger pairing.

import Foundation
import CoreBluetooth

let seconds = CommandLine.arguments.count > 1
    ? Int(CommandLine.arguments[1]) ?? 20 : 20

class Scanner: NSObject, CBCentralManagerDelegate {
    var seen: [String: (String, Int)] = [:]
    var mgr: CBCentralManager!

    func start() {
        mgr = CBCentralManager(delegate: self, queue: nil, options: nil)
    }

    func centralManagerDidUpdateState(_ central: CBCentralManager) {
        if central.state == .poweredOn {
            central.scanForPeripherals(withServices: nil, options: nil)
            print("scanning for \(seconds)s -- hold MO(1)+Q on the board NOW to pair")
        } else {
            print("Bluetooth not powered on (state=\(central.state.rawValue))")
            print("open System Settings > Bluetooth")
        }
    }

    func centralManager(_ central: CBCentralManager,
                        didDiscover peripheral: CBPeripheral,
                        advertisementData: [String: Any],
                        rssi RSSI: NSNumber) {
        let uuid = peripheral.identifier.uuidString
        let name = (peripheral.name
                    ?? (advertisementData[CBAdvertisementDataLocalNameKey] as? String)
                    ?? "(no name)")
        let rssi = RSSI.intValue
        // Only re-print first sighting per uuid.
        guard seen[uuid] == nil || seen[uuid]!.0 != name else { return }
        seen[uuid] = (name, rssi)
        let svcs = (advertisementData[CBAdvertisementDataServiceUUIDsKey] as? [CBUUID])?
            .map { $0.uuidString } ?? []
        let mfr = advertisementData[CBAdvertisementDataManufacturerDataKey] as? Data
        let mfrHex = mfr?.map { String(format: "%02x", $0) }.joined() ?? ""
        print(String(format: "[%4d dBm] %@  uuid=%@  mfr=%@  svcs=%@",
                     rssi, name, uuid, mfrHex, svcs.joined(separator: ",")))
    }
}

let s = Scanner()
s.start()
Thread.sleep(forTimeInterval: TimeInterval(seconds))
print("\n--- scan done, \(s.seen.count) unique device(s) ---")

File: scripts/ble-scan.py
#!/usr/bin/env python3
"""CoreBluetooth BLE advertisement scanner.

Listens for BLE advertisements for ~20s and prints every device seen, its
(name, RSSI, advertised services, manufacturer data). Used to check whether
the NEO65's CH582F radio is actually advertising when we tell it to pair.

Run inside: nix shell nixpkgs#python311Packages.pyobjc-core \
                    nixpkgs#python311Packages.pyobjc-framework-CoreBluetooth \
            -c python3 scripts/ble-scan.py
"""
import sys
import time
from Foundation import NSObject, NSRunLoop, NSDate
from CoreBluetooth import (
    CBCentralManager,
    CBManagerStatePoweredOn,
)

SECONDS = int(sys.argv[1]) if len(sys.argv) > 1 else 20

seen = {}  # uuid -> last snapshot


class Delegate(NSObject):
    def centralManagerDidUpdateState_(self, central):
        st = central.state()
        print(f"state={st} (poweredOn={CBManagerStatePoweredOn})")
        if st == CBManagerStatePoweredOn:
            central.scanForPeripheralsWithServices_options_(None, None)
            print(f"scanning for {SECONDS}s -- hold MO(1)+Q on the board NOW to pair")
        else:
            print("Bluetooth not powered on -- open System Settings > Bluetooth")

    def centralManager_didDiscoverPeripheral_advertisementData_RSSI_(
        self, central, peripheral, adv, rssi
    ):
        uuid = peripheral.identifier().UUIDString()
        name = peripheral.name() or adv.get("kCBAdvDataLocalName") or "(no name)"
        svcs = list(adv.get("kCBAdvDataServiceUUIDs", []) or [])
        mfr = adv.get("kCBAdvDataManufacturerData")
        mfr_hex = bytes(mfr).hex() if mfr else ""
        key = (uuid, str(name))
        prev = seen.get(key)
        # Only re-print if something changed or first sighting.
        snap = (str(name), int(rssi), mfr_hex, tuple(svcs))
        if prev != snap:
            seen[key] = snap
            print(f"[{int(rssi):>4} dBm] {name}  uuid={uuid}  mfr={mfr_hex}  svcs={svcs}")


def main():
    d = Delegate.alloc().init()
    mgr = CBCentralManager.alloc().initWithDelegate_queue_(d, None)
    loop = NSRunLoop.currentRunLoop()
    end = time.time() + SECONDS
    while time.time() < end:
        loop.runUntilDate_(NSDate.dateWithTimeIntervalSinceNow_(0.25))
    mgr.stopScan()
    print(f"\n--- scan done, {len(seen)} unique device(s) ---")


if __name__ == "__main__":
    main()
